File: solutions/zig/book-store/1/book_store.zig
const std = @import("std");

pub fn total(basket: []const u32) u32 {
    var books = [5]u32{ 0, 0, 0, 0, 0 };
    for (basket) |book| books[book - 1] += 1;
    std.mem.sort(u32, &books, {}, std.sort.asc(u32));

    var groups: [5]u32 = .{
        books[4] - books[3],
        books[3] - books[2],
        books[2] - books[1],
        books[1] - books[0],
        books[0],
    };

    const min = @min(groups[4], groups[2]);
    groups[4] -= min;
    groups[2] -= min;
    groups[3] += 2 * min;

    return 8 * (100 * groups[0] +
        95 * 2 * groups[1] +
        90 * 3 * groups[2] +
        80 * 4 * groups[3] +
        75 * 5 * groups[4]);
}

File: solutions/lua/book-store/1/book-store.lua
return {
	total = function(basket)
		local books = { 0, 0, 0, 0, 0 }
		for _, book in ipairs(basket) do
			books[book] = books[book] + 1
		end
		table.sort(books)
		local groups = {
			books[5] - books[4],
			books[4] - books[3],
			books[3] - books[2],
			books[2] - books[1],
			books[1],
		}
		local min = math.min(groups[5], groups[3])
		groups[5] = groups[5] - min
		groups[3] = groups[3] - min
		groups[4] = groups[4] + 2 * min

		local p1 = 100 * groups[1]
		local p2 = 95 * 2 * groups[2]
		local p3 = 90 * 3 * groups[3]
		local p4 = 80 * 4 * groups[4]
		local p5 = 75 * 5 * groups[5]
		return 8 * (p1 + p2 + p3 + p4 + p5)
	end,
}
